### PR TITLE
Fixes #20929 - Added fact name filtering on import

### PR DESCRIPTION
**Pull request by ShimShtein**
Added a setting that will filter out fact names, so they won't be recorded into `fact_names` and `fact_values` tables at all.

### CODE REVIEW

File: app/models/setting/provisioning.rb [hunk 1]
```diff
@@ -47,6 +47,7 @@ def self.default_settings
       self.set('manage_puppetca', N_("Foreman will automate certificate signing upon provision of new host"), true, N_('Manage PuppetCA')),
       self.set('ignore_puppet_facts_for_provisioning', N_("Stop updating IP address and MAC values from Puppet facts (affects all interfaces)"), false, N_('Ignore Puppet facts for provisioning')),
       self.set('ignored_interface_identifiers', N_("Ignore interfaces that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore interfaces with matching identifier')),
+      self.set('ignored_fact_names', N_("Ignore fact names that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore fact names with matching identifier')),
```
Comment (lzap) [inline app/models/setting/provisioning.rb]: Can you tell the user this includes puppet, rhsm and all facts in general?

File: app/services/fact_importer.rb [hunk 2]
```diff
@@ -116,13 +116,18 @@ def update_facts
...
   def normalize(facts)
+    fact_pairs = facts.map {|k, v| [k.to_s, v.to_s]}
+    # filter out matching fact names and empty values
+    fact_pairs.reject! {|k, v| v.empty? || k.match(ignored_fact_names)}
     # convert all structures to simple strings
-    facts = Hash[facts.map {|k, v| [k.to_s, v.to_s]}]
-    # and remove empty values
-    facts.keep_if { |k, v| v.present? }
+    Hash[fact_pairs]
   end
```
Comment (lzap) [inline app/services/fact_importer.rb]: I think it is a good opportunity to clean this method - this looks really ugly. Create pairs array, reject it, then create hash from it. I think creating copy of hash and then deleting filtered entries should be faster (leaving some buckets unused but faster).

    facts.select { |key, value| key.to_s.match(ignored_fact_names) }

This should do it, not tested tho.

File: db/migrate/20171122092336_add_ignored_fact_names_setting.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        'ignored_fact_names',
+        N_("Ignore fact names that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"),
+        Setting[:ignored_interface_identifiers],
+        N_('Ignore fact names with matching identifier'))) if Setting[:ignored_interface_identifiers]
+  end
+
+  def down
+    Setting.where(:name => :ignored_fact_names).destroy_all
+  end
+end
```
Comment (lzap) [inline db/migrate/20171122092336_add_ignored_fact_names_setting.rb]: One thing - this is a great opportunity to clean those facts from db. There are two choices:

A) Implicit - simply delete them in the migration. Can be dangerous for existing users.

B) Explicit - create a rake task and we can ask customer to run it (first show the facts, then confirm deletion). This will be safe enough for everybody.

It would be very nice to have the same rake task for NICs.

**Review by lzap — CHANGES_REQUESTED** [review]
We need this patch so hard, thank you so much :-)

File: app/models/setting/provisioning.rb (hunk 1 above)
Comment (ares) [reply app/models/setting/provisioning.rb]: The default value and second part of description seems wrong :-)

Comment (lzap) [reply app/models/setting/provisioning.rb]: Well it is correct, but I agree that the constant should be renamed in this case.

Comment (ares) [reply app/models/setting/provisioning.rb]: I didn't know there fact with names like macvtap*, perhaps fact names from newer facter. In such case, that's fine and just the constant should be named differently, it could be set to IGNORED_INTERFACES values though.

**Comment by ShimShtein** [comment]
Rebased, renamed, refactored and extended already existing `facts:clean` rake task

**Comment by lzap** [comment]
Marek pointed out on one important issue - fact names are different from interface names. 

```
interfaces => enp30s0,lo,redhat0,virbr0,virbr1,virbr2,virbr0_nic,virbr1_nic,virbr2_nic
ipaddress6_enp30s0 => fd5a:f687:b576:0:xxxx:f6ff:fe3e:aa52
ipaddress_enp30s0 => 192.168.x.13
macaddress_enp30s0 => 88:d7:f6:xx:xx:52
mtu_enp30s0 => 1500
netmask_enp30s0 => 255.255.255.0
network_enp30s0 => 192.168.1.0
```
I think we should be more explicit to prevent filtering out unwanted entries, therefore I suggest to create separate starting list. How about this:

```ruby
 IGNORED_FACTS =  IGNORED_INTERFACES.map{|i| "_#{i}"}
```

This will at least prefix all entries with underscore, which will match what we want.

File: test/unit/fact_cleaner_test.rb [hunk 4]
```diff
@@ -48,6 +48,19 @@ class FactCleanerTest < ActiveSupport::TestCase
     assert_include remaining_of(root_6_fact), root_6_fact
   end
 
+  test "it cleans excluded facts" do
+    cleaner.stubs(:excluded_facts).returns(['ignored*', '*bad'])
+    cleaner.stubs(:delete_orphaned_facts).returns(0)
```
Comment (lzap) [inline test/unit/fact_cleaner_test.rb]: This method (`delete_orphaned_facts`) needs a test.

File: app/models/setting/provisioning.rb [hunk 5]
```diff
@@ -48,6 +48,7 @@ def self.default_settings
       self.set('ignore_puppet_facts_for_provisioning', N_("Stop updating IP address and MAC values from Puppet facts (affects all interfaces)"), false, N_('Ignore Puppet facts for provisioning')),
       self.set('enable_orchestration_on_fact_import', N_("Enable host orchestration on puppet fact import. This could cause serious performance issues as the number of hosts increase"), false, N_('Enable orchestration on puppet fact import')),
       self.set('ignored_interface_identifiers', N_("Ignore interfaces that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore interfaces with matching identifier')),
+      self.set('excluded_facts', N_("Exclude pattern for all types of imported facts (rhsm, puppet e.t.c.). Those facts won't be stored in foreman's database. You can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Exclude pattern for facts stored in foreman')),
```
Comment (lzap) [inline app/models/setting/provisioning.rb]: Consider splitting this into separate default list with underscores (see my other comment).

File: app/services/fact_cleaner.rb [hunk 6]
```diff
@@ -2,11 +2,8 @@ class FactCleaner
   delegate :logger, :to => :Rails
 
   def clean!
-    to_delete = find_leaf_orphaned_facts
-    # we also delete all composes that either don't have any descendant or all of them are to be deleted
-    to_delete += find_compose_orphaned_facts(to_delete)
-    delete!(to_delete)
-    @deleted_count = to_delete.count
+    @deleted_count = delete_orphaned_facts
+    @deleted_count += delete_excluded_facts
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: This could be single line.

File: app/services/fact_cleaner.rb [hunk 7]
```diff
@@ -46,4 +51,29 @@ def delete!(to_delete)
     end
     logger.info "Cleanup of facts finished!"
   end
+
+  def find_excluded_facts
+    logger.info "Searching for facts that match excluded pattern..."
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Please `debug`. Frankly, I don't think this is much useful debug line anyway.

File: app/services/fact_cleaner.rb [hunk 8]
```diff
@@ -46,4 +51,29 @@ def delete!(to_delete)
...
+  def delete_excluded_facts
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.info "Removing values of excluded facts..."
+    FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.info "Removing excluded fact names..."
+    FactName.unscoped.where(:id => fact_names).delete_all
+    fact_names.count
+  end
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: This should be all `debug` lines.

File: db/migrate/20171122092336_add_ignored_fact_names_setting.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+        'excluded_facts',
+        N_("Ignore fact names that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"),
+        Setting[:ignored_interface_identifiers],
+        N_('Ignore fact names with matching identifier'))) if Setting[:ignored_interface_identifiers]
+  end
+
+  def down
+    Setting.where(:name => :excluded_facts).destroy_all
+  end
+end
```
Comment (lzap) [inline db/migrate/20171122092336_add_ignored_fact_names_setting.rb]: Is migration necessary? I thought settings are automatically created during boot time.

**Review by lzap — CHANGES_REQUESTED** [review]
Still have some comments.

File: test/unit/fact_cleaner_test.rb (hunk 4 above)
Comment (ShimShtein) [reply test/unit/fact_cleaner_test.rb]: `delete_orphaned_facts` was extracted out of `clean!`. All other tests are actually testing it.

File: app/services/fact_cleaner.rb [hunk 10]
```diff
@@ -39,11 +43,42 @@ def find_compose_orphaned_facts(found)
...
+      sql_pattern = exclude_pattern.gsub('*', '%')
+      # for facts that look like ignore_me::something
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', sql_pattern).pluck(:id)
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', sql_pattern + '::%').pluck(:id)
+      # for facts that look like common::ignore_me::something
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', '%::' + sql_pattern).pluck(:id)
+      fact_name_ids += FactName.unscoped.where('name LIKE ?', '%::' + sql_pattern + '::%').pluck(:id)
+    end
+    fact_name_ids
+  end
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Is this method new in the last commit? I haven't seen it. Anyway, this will generate *huge* SQL query and what's the worst - the server will need to run it for each individual fact upload. The cost for this consistency is way too high, I suggest to move this into rake task that can be run from cron or something once a week. This has potential to slow down fact import, we also haven't tested performance of this on MySQL.

File: app/services/fact_cleaner.rb [hunk 11]
```diff
@@ -39,11 +43,42 @@ def find_compose_orphaned_facts(found)
...
-    logger.info "Cleanup of facts finished!"
+    logger.debug "Cleanup of facts finished!"
+  end
+
+  def find_excluded_facts
+    logger.debug "Searching for facts that match excluded pattern..."
+    fact_name_ids = []
+
+    excluded_facts.each do |exclude_pattern|
+      sql_pattern = exclude_pattern.gsub('*', '%')
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Use faster `tr` instead of `gsub`. Also, this is vulnerable to SQL injection.

File: app/services/fact_cleaner.rb [hunk 12]
```diff
@@ -39,11 +43,42 @@ def find_compose_orphaned_facts(found)
...
+
+  def delete_excluded_facts
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.debug "Removing values of excluded facts..."
+    FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.debug "Removing excluded fact names..."
+    FactName.unscoped.where(:id => fact_names).delete_all
+    fact_names.count
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Seriously, we should not be doing regular cleaning for every request, but pull this to external process. This is way too many operations.

File: app/models/setting/provisioning.rb [hunk 13]
```diff
@@ -101,4 +107,9 @@ def self.map_pxe_kind
       yield pxe_kind, templates
     end
   end
+
+  def self.default_excluded_facts(ignored_interfaces = IGNORED_INTERFACES)
+    # By default ignore facts from both forms: "a::b::macvtap01::d::e" and "a => b => macvtap01 => ..."
+    ignored_interfaces.flat_map { |i| ["*::#{i}", i] }
+  end
```
Comment (lzap) [inline app/models/setting/provisioning.rb]: I don't understand, the list is enriched with `*::` here, and then in `find_excluded_facts` method, so is it added twice? I prefer just to have it in one place, the latter seems to be a better fit for this I think.

File: test/unit/fact_cleaner_test.rb [hunk 14]
```diff
@@ -48,6 +48,48 @@ class FactCleanerTest < ActiveSupport::TestCase
...
+    good_facts = [
+      FactoryBot.create(:fact_name, :name => 'not_ignored'),
+      FactoryBot.create(:fact_name, :name => 'not_ignored::something'),
+      FactoryBot.create(:fact_name, :name => 'common::not_ignored'),
+      FactoryBot.create(:fact_name, :name => 'common::not_ignored::something'),
+      FactoryBot.create(:fact_name, :name => 'ignored_shouldnt'),
+      FactoryBot.create(:fact_name, :name => 'ignored_shouldnt::something'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored_shouldnt'),
+      FactoryBot.create(:fact_name, :name => 'common::ignored_shouldnt::something')
+    ]
```
Comment (lzap) [inline test/unit/fact_cleaner_test.rb]: I must insist on real world facts, because (I haven't tested but) I believe your code does not work for real facts. Please include the following real facts I just captured from my server:

```
interfaces => em1,lo
ipaddress6_em1 => 2620:52:0:800:2247:ffff:ffff:95a0
ipaddress_em1 => 10.8.0.1
macaddress_em1 => 20:47:47:89:95:bb
mtu_em1 => 1500
netmask_em1 => 255.255.224.0
network_em1 => 10.8.0.0
dhcp_servers::em1 => 10.10.10.11
```

Again, I haven't tested myself, but I think your mathing code won't match most of those.

File: db/migrate/20171122092336_add_ignored_fact_names_setting.rb (hunk 9 above)
Comment (lzap) [reply db/migrate/20171122092336_add_ignored_fact_names_setting.rb]: What is the purpose of the migration?

**Review by lzap — DISMISSED** [review]
I have two big concerns. First, the deletion must not happen during fact upload, but via an external rake task (cron). This way, we can even make this more simple and simply iterate over all fact names via Ruby loop and filter them out using single regexp, you are trying to work in two "modes" (SQL and regexp) which makes the code too complex.

Second, you are not testing with real sample data. I think the code either does not match real world samples, or filters out what's not required. I think the pattern should be with underscore to match `macaddress_em1` (pattern `*_em1` or `dhcp-server::em1` (pattern `::em1`).

File: app/services/fact_cleaner.rb (hunk 10 above)
Comment (lzap) [reply app/services/fact_cleaner.rb]: I realized this method is only called from clean! method, so that is fine.

File: app/services/fact_cleaner.rb (hunk 11 above)
Comment (lzap) [reply app/services/fact_cleaner.rb]: Correction: No vulnerability :-)

File: app/services/fact_cleaner.rb (hunk 12 above)
Comment (lzap) [reply app/services/fact_cleaner.rb]: Disregard this comment.

File: app/models/setting.rb [hunk 15]
```diff
@@ -207,12 +207,14 @@ def self.create!(opts)
     end
   end
 
-  def self.regexp_expand_wildcard_string(string)
-    "\\A#{Regexp.escape(string).gsub('\*', '.*')}\\Z"
+  def self.regexp_expand_wildcard_string(string, options = {})
+    prefix = options[:prefix] || '\A'
+    suffix = options[:suffix] || '\Z'
+    "#{prefix}#{Regexp.escape(string).gsub('\*', '.*')}#{suffix}"
```
Comment (lzap) [inline app/models/setting.rb]: Nitpick: This could be simple string concat.

File: app/services/fact_cleaner.rb [hunk 16]
```diff
@@ -17,9 +13,17 @@ def deleted_count
 
   private
 
+  def delete_orphaned_facts
+    to_delete = find_leaf_orphaned_facts
+    # we also delete all composes that either don't have any descendant or all of them are to be deleted
+    to_delete += find_compose_orphaned_facts(to_delete)
+    delete!(to_delete)
+    to_delete.count
+  end
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: This should be:

    delete!(find_compose_orphaned_facts(find_leaf_orphaned_facts))

as the method returns number of rows deleted.

File: app/services/fact_cleaner.rb [hunk 17]
```diff
@@ -39,11 +43,44 @@ def find_compose_orphaned_facts(found)
...
-    logger.info "Cleanup of facts finished!"
+    logger.debug "Cleanup of facts finished!"
+  end
+
+  def find_excluded_facts
+    logger.debug "Searching for facts that match excluded pattern..."
+    fact_name_ids = []
+
+    excluded_facts.each do |exclude_pattern|
+      sql_pattern = exclude_pattern.gsub('*', '%')
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Method `tr` is still much faster.

File: app/services/fact_cleaner.rb [hunk 18]
```diff
@@ -39,11 +43,44 @@ def find_compose_orphaned_facts(found)
...
+
+  def delete_excluded_facts
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.debug "Removing values of excluded facts..."
+    FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.debug "Removing excluded fact names..."
+    FactName.unscoped.where(:id => fact_names).delete_all
+    fact_names.count
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Would be better to actually return number of deleted rows as it gets logged. Best thing would be to add debug lines for both number of rows planned to delete and number which actually was deleted.

File: test/unit/fact_cleaner_test.rb [hunk 19]
```diff
@@ -48,6 +48,49 @@ class FactCleanerTest < ActiveSupport::TestCase
...
   end
 
+  test "it cleans excluded facts" do
+    cleaner.stubs(:excluded_facts).returns(['ignored*', '*bad'])
+    cleaner.stubs(:delete_orphaned_facts).returns(0)
+    good_fact_name = FactoryBot.create(:fact_name, :name => 'good_fact')
+    ignored_fact_name = FactoryBot.create(:fact_name, :name => 'ignored01')
+    bad_fact_name = FactoryBot.create(:fact_name, :name => 'empty_bad')
+    FactoryBot.create(:fact_value, :fact_name => good_fact_name)
+    FactoryBot.create(:fact_value, :fact_name => ignored_fact_name)
```
Comment (lzap) [inline test/unit/fact_cleaner_test.rb:63]: I still do not see testing of "macaddress_em1" kind of fact. Frankly, I was hoping for real-world interface names rather than these artificial ones :-)

File: test/unit/fact_importer_test.rb [hunk 20]
```diff
@@ -55,6 +55,73 @@ def fact_name_class
...
+        'something::ignored_fact' => 'will_not_show',
+        'something::ignored_fact::something_else' => 'will_not_show',
+        'fact_bad::something' => 'will_not_show',
+        'something::fact_bad' => 'will_not_show',
+        'something::fact_bad::something_else' => 'will_not_show',
+        'filter::something' => 'will_not_show',
+        'something::filter' => 'will_not_show',
+        'something::filter::something_else' => 'will_not_show',
+        'something::something_else' => 'hello'
+      }
```
Comment (lzap) [inline test/unit/fact_importer_test.rb]: Can you also add edge cases: "ignored_fact" and "fact_bad" (without any `::` characters). I am not sure if RHSM generates it, but it's good to be safe.

File: test/unit/fact_importer_test.rb [hunk 21]
```diff
@@ -55,6 +55,73 @@ def fact_name_class
...
+
+      importer = FactImporter.new(nil, facts)
+      actual_facts = importer.send(:facts)
+
+      assert_not_nil actual_facts['something::something_else']
+      assert_equal 1, actual_facts.count
```
Comment (lzap) [inline test/unit/fact_importer_test.rb]: Can you also assert that all passed facts have "hello" in its value? Just to be sure it filtered it correctly :-) In all tests please.

File: test/unit/fact_importer_test.rb [hunk 22]
```diff
@@ -55,6 +55,73 @@ def fact_name_class
...
+        # facter < 3.0 style facts: macaddress_virbr0
+        'ignored_fact_something' => 'will_not_show',
+        'something_ignored_fact' => 'will_not_show',
+        'something_ignored_fact_something_else' => 'will_not_show',
+        'something_fact_bad' => 'will_not_show',
+        'something_filter' => 'will_not_show',
+        'something_filter_something_else' => 'hello',
+        'fact_bad_something' => 'hello',
+        'filter_something' => 'hello'
+      }
```
Comment (lzap) [inline test/unit/fact_importer_test.rb]: Also edge cases: "ignore" and "_bad" and "filter" names should all be filtered. Although it might not make sense, users could try to filter some facts with this.

File: test/unit/structured_fact_importer_test.rb [hunk 23]
```diff
@@ -78,6 +78,40 @@ class StructuredFactImporterTest < ActiveSupport::TestCase
...
+          },
+          :good_subtree => {
+            :good_key => 'good_value'
+          }
+        },
+        :empty_ancestor => {
+          :ignored_key => 'some_val'
+        },
+        :ignored_value => 'some_val'
+      }
```
Comment (lzap) [inline test/unit/structured_fact_importer_test.rb]: Honestly, I'd appreciate common testing value style across all those tests - either "yes" or "no" or whatever values so these can be clearly identified :-)

I think it's very important to make sure both cleaning task and filtering here works the same - can you move the fact testing data (all three types) into some helper and reuse it in both tests? This way we can be sure if someone makes change in the future, failing test will catch the other workflow.

**Review by lzap — CHANGES_REQUESTED** [review]
Overall no blockers but I don't want to rush this. Can you improve tests, I am still missing one and I have a proposal to share testing data across both two workflows to prevent regressions.

**Comment by ShimShtein** [comment]
Finally! The tests are green. Had to give up cleaner's DB code, there is no way to escape `LIKE` query so it will work on all databases.

File: app/services/fact_cleaner.rb [hunk 24]
```diff
@@ -39,11 +43,38 @@ def find_compose_orphaned_facts(found)
...
     end
-    logger.info "Cleanup of facts finished!"
+    logger.debug "Cleanup of facts finished!"
+  end
+
+  def find_excluded_facts
+    logger.debug "Searching for facts that match excluded pattern..."
+    fact_name_ids = []
+
+    excluded_facts_regex = FactImporter.new(nil, {}).send(:excluded_facts)
```
Comment (lzap) [inline app/services/fact_cleaner.rb]: Uh. Oh.

File: app/services/fact_cleaner.rb [hunk 25]
```diff
@@ -39,11 +43,38 @@ def find_compose_orphaned_facts(found)
...
+    fact_names = find_excluded_facts
+
+    # delete related fact values first
+    logger.debug "Removing values of excluded facts..."
+    deleted_values = FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
+    logger.debug "Removed #{deleted_values} associated fact value records."
+    logger.debug "Removing excluded fact names..."
+    deleted_names = FactName.unscoped.where(:id => fact_names).delete_all
+    logger.debug "Removed #{deleted_names} fact name records."
+    deleted_names
```
Comment (lzap) [inline app/services/fact_cleaner.rb:78]: I am usually against debug messages like "I am doing something..." the "Removed" lines are enough to tell you it has been processed.

File: app/services/fact_importer.rb [hunk 26]
```diff
@@ -187,4 +191,14 @@ def ensure_no_active_transaction
...
+
+  def excluded_facts
+    @excluded_facts ||= Setting.convert_array_to_regexp(
+      Setting[:excluded_facts],
+      {
+        :prefix => '(\A|.*::|.*_)',
+        :suffix => '(\Z|::.*)'
+      }
+    )
+  end
```
Comment (lzap) [inline app/services/fact_importer.rb:207]: Okay, this can be moved to settings, why not? The other line with send really hurts :-)

File: test/fact_importer_test_helper.rb [hunk 27]
```diff
@@ -34,3 +34,119 @@ def instance_stubs
...
+        'something::ignored_fact::something_else' => 'will_not_show',
+        'fact_bad::something' => 'will_not_show',
+        'something::fact_bad' => 'will_not_show',
+        'something::fact_bad::something_else' => 'will_not_show',
+        'filter::something' => 'will_not_show',
+        'something::filter' => 'will_not_show',
+        'something::filter::something_else' => 'will_not_show'
+      }
+    end
+  end
```
Comment (lzap) [inline test/fact_importer_test_helper.rb:86]: One could say this belongs to Katello, but I really think it should be in core. Thanks.

File: app/models/setting/provisioning.rb [hunk 28]
```diff
@@ -101,4 +107,8 @@ def self.map_pxe_kind
       yield pxe_kind, templates
     end
   end
+
+  def self.default_excluded_facts(ignored_interfaces = IGNORED_INTERFACES)
+    ignored_interfaces
+  end
```
Comment (lzap) [inline app/models/setting/provisioning.rb:113]: Can you optionally add couple of tests to test IGNORED_INTERFACES constant actually? We are not sure we ship with default value that just works. A single test with different names "Ipaddress_vnet1" should do it "Ipaddress:vnet1".

**Review by lzap — CHANGES_REQUESTED** [review]
One ugly line and few nitpicks.

File: app/services/fact_cleaner.rb (hunk 25 above)
Comment (ShimShtein) [reply app/services/fact_cleaner.rb:78]: It really helps to identify slow queries - you will see the "starting.." message, but not the "finished" one.

File: app/services/fact_importer.rb (hunk 26 above)
Comment (ShimShtein) [reply app/services/fact_importer.rb:207]: Since we don't have a special settings class, I prefer keeping it with the importer. Refactored it to a class method though.

File: test/fact_importer_test_helper.rb (hunk 27 above)
Comment (ShimShtein) [reply test/fact_importer_test_helper.rb:86]: Katello is only an example of this kind of facts. Since this notation is supported by core, I think it should be tested in core.

File: app/services/fact_cleaner.rb (hunk 25 above)
Comment (lzap) [reply app/services/fact_cleaner.rb:78]: Fair enough, although AR time is reported for each request by Rails framework and then there is SQL server tracking, but I buy that. We miss TRACE level so much.

File: app/services/fact_importer.rb (hunk 26 above)
Comment (lzap) [reply app/services/fact_importer.rb:207]: As long as the send method is gone, okay :-)

**Review by lzap — APPROVED** [review]
Thanks, merging.

**Comment by lzap** [comment]
Katello test failures seem irrelevant.
